File: utils/go/union_find.go
package utils

// Union-Find
type unionFind struct {
	parent []int
	rank   []int
	count  int // number of connected components
}

func newUnionFind(n int) *unionFind {
	parent := make([]int, n)
	rank := make([]int, n)

	for x := range n {
		parent[x] = x
		rank[x] = 0
	}

	return &unionFind{
		parent: parent,
		rank:   rank,
		count:  n,
	}
}

func (uf *unionFind) find(x int) int { // path compression
	if uf.parent[x] != x {
		uf.parent[x] = uf.find(uf.parent[x])
	}
	return uf.parent[x]
}

func (uf *unionFind) union(x, y int) { // union with rank
	x = uf.find(x)
	y = uf.find(y)
	if x == y {
		return
	}

	// Ensure rank(x) >= rank(y)
	if uf.rank[x] < uf.rank[y] {
		x, y = y, x
	}

	// Marge y into x
	uf.parent[y] = x
	if uf.rank[x] == uf.rank[y] {
		uf.rank[x]++
	}

	uf.count--
}

File: utils/cpp/union_find.cpp
#include <numeric>
#include <vector>

using namespace std;

class UnionFind {
  int n;
  vector<int> parents;
  vector<int> ranks;

  UnionFind(int n) : n(n), parents(n), ranks(n) {  //
    iota(parents.begin(), parents.end(), 0);
  }

  int find(int x) {
    if (parents[x] != x) parents[x] = find(parents[x]);
    return parents[x];
  }

  void unite(int x, int y) {
    x = find(x);
    y = find(y);
    if (x == y) return;

    // Ensure rank(x) >= rank(y)
    if (ranks[x] < ranks[y]) swap(x, y);

    // Merge y into x
    if (ranks[x] == ranks[y]) ++ranks[x];
    parents[y] = x;
  }
};
